File: src/files/SystemFileManager.java
/*
 * Copyright (c) 2015. This code is a LogosProg property. All Rights Reserved.
 */

package files;

import java.io.*;

/**
 * Created by forando on 14.06.15.<br>
 * This class provides basic operations with files that are
 * used by each module/application. That's why they are called
 * system files.
 */
public class SystemFileManager {

    /**
     * A name to be given to a new file
     */
    protected String fileName;
    /**
     * The complete path to root directory.
     */
    protected String rootDir;
    /**
     * Optional. The subdirectory to root directory.
     *          if it's given, then it will be included in the fileName path
     */
    protected String subDir;

    public SystemFileManager(String fileName, String rootDir, String... subDir) throws IOException{
        if (fileName == null || rootDir == null || rootDir.isEmpty() || fileName.isEmpty())
            throw new IOException("You must pass in fileName and rootDir");
        this.fileName = fileName;
        this.rootDir = rootDir;
        if (null != subDir && subDir.length > 0){
            //next: proceed with multiple subdirectories
            this.subDir = subDir[0];
        }
    }

    public String getFileName(){
        return fileName;
    }


    public boolean fileExists(){

        if (null != subDir){
            return (new File(getFilePath())).exists();
        }else{
            return (new File(getFilePath())).exists();
        }
    }

    /**
     * This method constructs whole path to desired system file.<br>
     *     This is where you can change the default location of all
     *     files of your application.
     *     if it's given, then it will be included in the fileName path
     * @return Constructed path to the requested file
     */
    protected String getFilePath(){
        if (null != subDir){
            return rootDir + File.separator + subDir + File.separator + fileName;
        }else{
            return rootDir + File.separator + fileName;
        }

    }

    /**
     * Creates parent DIR if it does not exist and than creates
     * an empty file
     * @return True - if operation is successful
     */
    public boolean createEmptyFile(){

        boolean result = false;
        boolean dirCreated = true;

        File f = new File(getFilePath());

        /*
        Checking if directory for files already exists
         */
        if (!f.getParentFile().exists()) dirCreated = f.getParentFile().mkdirs();

        try {
            result = dirCreated && f.createNewFile();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return result;
    }


    public boolean deleteDefaultFile(){
        boolean result = false;

        File f = new File(getFilePath());
        try{
            if (f.delete()){
                result = true;
            }
        }catch (Exception ex){
            ex.printStackTrace();
        }

        return result;
    }

    /**
     * This method copies default file from internal app folders to specific
     * project folder.
     * @return True - if operation is successful.
     */
    public boolean generateDefaultFile(){

        boolean result = false;

        InputStream is = null;
        OutputStream os = null;
        String path;

        if (null != subDir) {
            path = getFilePath();
        }else{
            path = getFilePath();
        }

        if (!createEmptyFile()) return false;

        try {

            is = SystemFileManager.class.getResourceAsStream(fileName);
            os = new FileOutputStream(path);
            byte[] buffer = new byte[1024];
            int length;
            while ((length = is.read(buffer)) > 0) {
                os.write(buffer, 0, length);
            }
            result = true;
        }catch(IOException ex){
            ex.printStackTrace();
        }finally {
            if (is != null) {
                try {
                    is.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            if (os != null) {
                try {
                    os.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        return result;
    }
}

File: src/main/kotlin/com/logosprog/mainutils/files/managers/SystemFileManager.kt
/*
 * Copyright (c) 2016. This code is a LogosProg property. All Rights Reserved.
 */

/*
 * Copyright (c) 2016. This code is a LogosProg property. All Rights Reserved.
 */

package com.logosprog.mainutils.files.managers

import java.io.*

/**
 * Provides basic operations with files.
 *
 * @param fileName A name to be given to a new file
 * @param rootDir The complete path to a root directory.
 * @param subDir Optional. The subdirectory to root directory.
 * @throws [IOException] if either [fileName] or [rootDir] is empty
 */
open class SystemFileManager(val fileName: String, private val rootDir: String, private val subDir: String? = null){

    init {
        if (rootDir.isEmpty() || fileName.isEmpty())
            throw IOException("You must pass in fileName and rootDir")
    }

    fun fileExists(): Boolean = File(getFilePath()).exists()

    /**
     * This method constructs whole path to desired system file.
     * This is where you can change the default location of all
     * files of your application.
     * if it's given, then it will be included in the fileName path
     * @return Constructed path to the requested file
     */
    protected fun getFilePath(): String {
        return if (null != subDir) {
            rootDir + File.separator + subDir + File.separator + fileName
        } else {
            rootDir + File.separator + fileName
        }
    }

    /**
     * Creates parent DIR if it does not exist and than creates
     * an empty file
     * @return True - if operation was successful.
     */
    fun createEmptyFile(): Boolean {
        val f = File(getFilePath())
        var dirCreated = true
        var newFileCreated = false

        try {
            if (!f.parentFile.exists())
                dirCreated = f.parentFile.mkdirs()
            newFileCreated = f.createNewFile()
        } catch (ex: Exception) {
            ex.printStackTrace()
        }

        return dirCreated && newFileCreated
    }

    /**
     * Deletes the predefined file.
     * @return True - if operation was successful.
     */
    fun deleteDefaultFile(): Boolean = File(getFilePath()).delete()

    /**
     * This method copies default file from internal resource folder to a specific
     * project folder.
     * @return True - if operation was successful.
     */
    fun generateDefaultFile(): Boolean {
        val path = getFilePath()
        val inputStream: InputStream = this.javaClass.classLoader.getResourceAsStream(fileName) ?: return false
        if (!createEmptyFile()){
            inputStream.close()
            return false
        }
        try {
            FileOutputStream(path).use {
                val buffer = ByteArray(1024)
                var length = 0

                while (
                        inputStream
                                .read(buffer)
                                .let {
                                    length = it
                                    it > 0
                                }
                ) {
                    it.write(buffer, 0, length)
                }
            }
            return true
        } catch (ex: Exception) {
            ex.printStackTrace()
            return false
        }finally {
            inputStream.close()
        }

        try {
            inputStream.close()
        }catch (ignored: IOException){}

        return true
    }
}

